proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,237 @@
+start of the epoch. The inputs, vote_accounts/stake_account, to the rewards
+calculation will be identical.
+
+To avoid this repeated computation, the signature of the computation requests
+is calculated.
+
+```
+calculation_signature = hash(epoch_number, [&stake_accounts], [&vote_accounts], [&stake_delegations])
+```
+
+Computation requests with same signature are discarded. In the above example,
+there are no stake/vote accounts changes between slot `N` and `N+2`, and the
+second computation request will be discarded.
+
+The reward computation phase lasts for `N` blocks. When reaching block height
+`N` after the start of the `reward calculation phase`, the bank will mark it
+the end the `reward calculation phase`.
+
+At this point, the reward computation result should be available. In case that
+the reward computation result is not available yet at the `N_th_` block-height,
+the bank runtime will halt and wait until the computation is finished, before
+processing any blocks with height greater than `N`.
+
+At the end of the reward computation phase, a set of sysvar accounts will be
+updated - `EpochRewardProof` and `RewardReserve`.
+
+`EpochRewardProof` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardProof` is an array of entries,
+```
+[(epoch, (reward_amount, root_hash))]
+```
+, where each entry corresponds to the reward for
+one particular epoch. The maximum depth of the history are the same as the
+depth of stake history, which is 512. The `root_hash` of the merkle tree for
+all rewards are optional. It is for potential future works to distribute
+rewards as transactions.
+
+When reaching the end of reward computation phase, `EpochRewardProof` are
+updated with the reward computation result. A new entry will be pushed into the
+array. If the length of the history array reaches the maximum depth, the oldest
+entry will popped.
+
+```
+struct RewardHistoryEntry {
+   total: u64,                  // total rewards for this epoch
+   distributed: u64,            // already distributed reward amount
+   root_hash: Option<Hash>,     // root-hash of reward merkle tree
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+At the end of reward computation phase, a set of `EpochRewardsReserve` sysvar
+accounts will be updated too. Each `RewardsReserve` records the reward balance
+to be distributed for a particular block in `Reward Distribution` phase, which
+captures a particular subset of stakes rewards to be paid at the `index` block in
+the reward distribution phase. The complete subset of stake rewards forms a
+partition of the all the stake rewards.
+
+```
+struct EpochRewardReserve {
+   balance: u64,
+   index: u64,
+   hash: Hash
+}
+```
+
+`balance` is total amount of rewards to be distributed in this block.
+`index` is the reward distribution block hight offset from reward computation end block.
```

Comment:
```suggestion
`index` is the reward distribution block height offset from reward computation end block.
```

also since `index’ is a `u64`, using the absolute block height instead of an offset would probably be easier on users that want to read the value